### ISSUE TITLE
Document the flutter_test dimensions and describe tricks for testing other dimensions

### ISSUE THREAD

**Issue by robertoscaramuzzi** [flutter/flutter#12994]
Currently the size is fixed at 800 x 600, which makes it impossible to test behavior that happens only in portrait mode, as well as other size-dependent behavior.

**Comment by tvolkert** [comment]
Potential options:

1. Support flags to `flutter_tester` that affect the screen resolution in [test_runner.cc](https://github.com/flutter/engine/blob/1f0c959189dc1057556b0a28f486ef1634351012/shell/testing/test_runner.cc#L20-L22)
2. Make [`AutomatedTestWidgetsFlutterBinding`](https://github.com/flutter/flutter/blob/9be85b6b4bf41df834c676d9feac20c0657ac812/packages/flutter_test/lib/src/binding.dart#L522) override `createViewConfiguration()` ([like `LiveTestWidgetsFlutterBinding` does](https://github.com/flutter/flutter/blob/9be85b6b4bf41df834c676d9feac20c0657ac812/packages/flutter_test/lib/src/binding.dart#L916)) to look for some hook to override the default resolution.

**Comment by tvolkert** [comment]
Actually it looks like:

* #<span></span>1 above is required.
* That will make `AutomatedTestWidgetsFlutterBinding` automatically pick up the desired resolution.
* We would need to update `LiveTestWidgetsFlutterBinding` to provide a hook whereby the developer could specify the fixed resolution in which they want to run.  Otherwise, even though the real test was run in a resolution specified by the developer, a `flutter run` invocation of that test would yield a different resolution.

**Comment by Hixie** [comment]
Just wrap the widget under test in a SizedBox (contained in a Center) with the given size. Optionally add a MediaQuery in there too if that matters.

**Comment by robertoscaramuzzi** [comment]
Thanks, Hixie. That seems to work, but it's entirely non-obvious. It would be good to provide a better API.

**Comment by Hixie** [comment]
We should definitely document it, maybe provide some arguments to pumpWidget to do it for you.

**Comment by alanrussian** [comment]
IIRC the current default size isn't documented anywhere. That might be helpful to document too.

**Comment by alanrussian** [comment]
Not sure how much this is related but it would be useful if we could take full-page screenshots of entire widgets for screenshot tests. For example: if you have a `ListView` that extends beyond the default viewport, the screenshot would capture the entire `ListView`. Also, if you have a widget less than the default viewport, it could capture an image that's just that widget's size.

**Comment by tvolkert** [comment]
At least the second use case is possible by inserting a `RepaintBoundary` in the hierarchy where you want to cap the screenshot.

**Comment by Hixie** [comment]
ListViews don't go beyond the viewport. That's sort of the entire point of a ListView. :-) They don't even reify the widgets that aren't visible.

You could always create a RepaintBoundary around a SizedBox that is the size you want, though, and put the ListView inside that.

**Comment by tvolkert** [comment]
@Hixie, @paolosoares reports that he'd like to not just change the size of the produced image -- he'd like to change the constraints so that the widget lays out differently.

**Comment by Hixie** [comment]
@alanrussian ok that sounds like a separate issue. Please file a separate bug for that.
@tvolkert sure. my earlier comment describes how to do that.

**Comment by zoechi** [comment]
This worked for me
```dart
setUp(() {
  WidgetsBinding.instance.renderView.configuration =  new TestViewConfiguration(size: const Size(1200.0, 1980.0));
});
```

**Comment by zoechi** [comment]
Related
- #10307
- #6380

**Comment by zoechi** [comment]
See also https://stackoverflow.com/questions/53706569/how-to-test-flutter-widgets-on-different-screen-sizes

**Comment by Hixie** [comment]
We should probably provide a "TestSize" or "TestScreen" widget in flutter_test that takes a Size and a key and creates the structure: MediaQuery -> OverflowBox (using the size) -> SizedBox (using the size) -> RepaintBoundary (using the key).

**Comment by matejwamf** [comment]
> This worked for me
> 
> ```dart
> setUp(() {
>   WidgetsBinding.instance.renderView.configuration =  new TestViewConfiguration(size: const Size(1200.0, 1980.0));
> });
> ```

This could be better documented in the test class. I found the `TestViewConfiguration` inside test package, but there were no info on how to use it. Thank you.

**Comment by aanelson** [comment]
I fixed it by setting the pixelRatio to 1 and the physical size because it is ultimately the MediaQuery.from where the size is being set to default

e.g.
`    
await tester.binding.setSurfaceSize(size);
tester.binding.window.physicalSizeTestValue = size; 
tester.binding.window.devicePixelRatioTestValue = 1;
`

**Comment by mikeesouth** [comment]
> I fixed it by setting the pixelRatio to 1 and the physical size because it is ultimately the MediaQuery.from where the size is being set to default
> 
> e.g.
> 
> > await tester.binding.setSurfaceSize(size);
> > tester.binding.window.physicalSizeTestValue = size;
> > tester.binding.window.devicePixelRatioTestValue = 1;

Setting tester.binding.window.physicalSizeTestValue and devicePixelRatioTestValue helped me so that I can see the actual size of the rendered area even if I run the test on a running emulator.

I do however still feel that the support for this testing is lacking more "official" support. I have a test that is a rather complex widget. The widget test fails with a render overflow on 22 pixels. If I create an emulator with the same resolution (800x600) it works as expected (no overflow). The strangest thing is that if I run my test with `flutter run test/my_test.dart` in a huge emulator I see the smaller 800x600 area is the only thing being rendered - without overflow and the test passes.

So, the test passes when running using `flutter run test/my_test.dart` but it fails when running via flutter_test. I have checked that all size calculations, media query sizes, paddings, viewinsets and other stuff is the same. It just overflows when running `flutter test` and it doesn't overflow when running use `flutter run`. Strange.

I have tried not setting the surface size (i.e. defaulting to 800x600) and setting it using the various methods described in this thread - nothing helps. The widget does not overflow if I set it to e.g. 1000x800 even when run with `flutter test`.

In other words - it seems like `flutter test` still has some padding or other rendering logic that isn't reflected in `flutter run` which is unfortunate.

I can't really share my code in it's current state but any ideas on how I can fix this is appreciated.

**Comment by acoutts** [comment]
Is there any solution to this? I can't get my widget test to run because of a phantom overflow somewhere.

I tried adjusting MediaQuery and this:
```dart
        final size = Size(414, 896);
        await binding.setSurfaceSize(size);
        binding.window.physicalSizeTestValue = size;
        tester.binding.window.devicePixelRatioTestValue = 1;
```

But I can't fix this overflow. Those dimensions are the same as the iphone simulator and it runs fine in that. How can we fix widget test sizing?

**Comment by jamesderlin** [comment]
> Just wrap the widget under test in a SizedBox (contained in a Center) with the given size. Optionally add a MediaQuery in there too if that matters.

For simulating screen rotation, it seems that the `SizedBox` + `MediaQuery` approach might not behave quite like an actual screen rotation on a device.  I've observed an actual screen rotation triggering `Layer.attach` without triggering `Layer.addToScene`, but changing the dimensions of a `SizedBox` seems to trigger both.

**Comment by markusaksli-nc** [comment]
flutter/website#2005 is relevant as well here

**Comment by juanmendez** [comment]
> This worked for me
> 
> ```dart
> setUp(() {
>   WidgetsBinding.instance.renderView.configuration =  new TestViewConfiguration(size: const Size(1200.0, 1980.0));
> });
> ```

it was just magic when dealing with an app made for a tablet. thank you!

**Comment by dnfield** [comment]
The correct way to do this is to call `setSurfaceSize` on the binding.

E.g.

```dart
tester.binding.setSurfaceSize(size);
```
